compiler/damlc/tests/src/VisualTest.hs

```diff
@@ -0,0 +1,111 @@
+-- Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+module VisualTest
+   ( main
+   ) where
+import DA.Cli.Visual
+import Data.Either
+import qualified Test.Tasty.Extended as Tasty
+import qualified Test.Tasty.HUnit    as Tasty
+import qualified DA.Pretty as DAP
+import qualified Development.IDE.Types.Location as D
+import qualified DA.Daml.LF.Ast as LF
+import qualified Development.IDE.Core.API as API
+import qualified Development.IDE.Core.Rules.Daml as API
+import qualified Data.NameMap as NM
+import qualified Control.Monad.Reader   as Reader
+import Development.IDE.Core.API.Testing
+import System.Environment.Blank (setEnv)
+import Control.Monad.Except
+
+data ExpectedChoices = ExpectedChoices
+    { _cName :: String
+    , _consuming :: Bool
+    } deriving (Eq, Show )
+data TemplateProp = TemplateProp
+    { _choices :: [ExpectedChoices]
+    , _action :: Int
+    } deriving (Eq, Show)
+
+main :: IO ()
+main = do
+    setEnv "TASTY_NUM_THREADS" "1" True
+    Tasty.deterministicMain visualTests
+
+visualTests :: Tasty.TestTree
+visualTests =
+    Tasty.testGroup "Visual tests using Shake API" [ visualDamlTests ]
+
+testCase :: Tasty.TestName -> ShakeTest () -> Tasty.TestTree
+testCase testName test =
+    Tasty.testCase testName $ do
+        res <- runShakeTest Nothing test
+        Tasty.assertBool ("Shake test resulted in an error: " ++ show res) $ isRight res
+
+templateChoicesToProps :: TemplateChoices -> TemplateProp
+templateChoicesToProps tca  = TemplateProp choicesInTpl actl
+    where choicesInTpl = map (\ca -> ExpectedChoices ( DAP.renderPretty $ choiceName ca) (choiceConsuming ca)) (choiceAndActions tca)
+          actl = sum $ map (length . actions ) (choiceAndActions tca)
+
+graphTest :: LF.World -> LF.Package -> [TemplateProp] -> Either (String, String) ()
+graphTest wrld lfPkg tplProps =
+    if tplProps == map templateChoicesToProps tplPropsActual
+        then Right ()
+        else Left (show tplProps, show $ map templateChoicesToProps tplPropsActual)
+    where tplPropsActual = concatMap (moduleAndTemplates wrld) (NM.toList $ LF.packageModules lfPkg)
+
+expectedPoperties :: D.NormalizedFilePath -> [TemplateProp] -> ShakeTest ()
+expectedPoperties damlFilePath tplProps = do
+    ideState <- ShakeTest $ Reader.asks steService
+    mbDalf <- liftIO $ API.runAction ideState (API.getDalf damlFilePath)
+    case mbDalf of
+        Just lfPkg -> do
+            wrld <- Reader.liftIO $ API.runAction ideState (API.worldForFile damlFilePath)
+            case graphTest wrld lfPkg tplProps of
+                Right _ -> pure ()
+                Left (expected , actual) -> throwError (ExpectedNoMisMatch expected actual)
+        Nothing -> throwError (ExpectedNoErrors [])
```

Comment:
`ExpectedNoErrors []` isn’t right here: The list you are passing to it should be the lists of errors you got but by passing the empty lists you are indicating that you got no errors which doesn’t make sense. I would suggest that you use `expectNoErrors` and then just turn the pattern match failure here into an exception.

Reply:
Now returning back the expected and the actual, in this case, the actual will be empty